File: examples/c++/solid_mechanics_cohesive_model/my_first_test/my_first_test.cc
/**
 * Copyright (©) 2012-2023 EPFL (Ecole Polytechnique Fédérale de Lausanne)
 * Laboratory (LSMS - Laboratoire de Simulation en Mécanique des Solides)
 *
 * This file is part of Akantu
 *
 * Akantu is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * Akantu is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Akantu. If not, see <http://www.gnu.org/licenses/>.
 */

/* -------------------------------------------------------------------------- */
#include "element_group.hh"
#include "mesh_iterators.hh"
#include "non_linear_solver.hh"
#include "solid_mechanics_model_cohesive.hh"
/* -------------------------------------------------------------------------- */
#include <iostream>
/* -------------------------------------------------------------------------- */

using namespace akantu;

/* -------------------------------------------------------------------------- */
int main(int argc, char * argv[]) {
  initialize("material.dat", argc, argv);

  const Int spatial_dimension = 2;
  const Int max_steps = 1;

  Mesh mesh(spatial_dimension);
  mesh.read("triangle.msh");

  SolidMechanicsModelCohesive model(mesh);
  model.getElementInserter().setLimit(_x, -0.05, 0.05);

  /// model initialization
  model.initFull(_analysis_method = _static
                 ,_is_extrinsic = false
                 );

  auto & solver = model.getNonLinearSolver();
//  solver.set("convergence_type", SolveConvergenceCriteria::_residual);
  solver.set("convergence_type", SolveConvergenceCriteria::_solution);
  solver.set("max_iterations", 2);

  model.applyBC(BC::Dirichlet::FixedValue(0., _x), "left");
  model.applyBC(BC::Dirichlet::FixedValue(0., _y), "point");

  model.setBaseName("cohesive");
  model.addDumpFieldVector("displacement");
//  model.addDumpField("velocity");
//  model.addDumpField("acceleration");
  model.addDumpField("stress");
  model.addDumpField("grad_u");
  model.addDumpField("external_force");
  model.addDumpField("internal_force");
  model.dump();


  Real increment = 0.01;

  /// Main loop
  for (Int s = 1; s <= max_steps; ++s) {
    model.applyBC(BC::Dirichlet::IncrementValue(increment, _x), "right");
    // debug::setDebugLevel(dblDump);
    debug::setDebugLevel(dblInfo);
    model.solveStep();
    debug::setDebugLevel(dblWarning);

    if (s % 1 == 0) {
      model.dump();
      std::cout << "passing step " << s << "/" << max_steps << std::endl;
    }
  }

  Real E = model.getEnergy("potential");

  std::cout << "Elastic energy : " << E << std::endl;

//  if (Ed < Edt * 0.999 || Ed > Edt * 1.001 || std::isnan(Ed)) {
//    std::cout << "The dissipated energy is incorrect" << std::endl;
//    return EXIT_FAILURE;
//  }

  finalize();

  return EXIT_SUCCESS;
}

File: examples/c++/solid_mechanics_cohesive_model/my_first_test/my_first_test.py
#!/usr/bin/env python3
__copyright__ = (
    "Copyright (©) 2019-2023 EPFL (Ecole Polytechnique Fédérale de Lausanne)"
    "Laboratory (LSMS - Laboratoire de Simulation en Mécanique des Solides)"
)
__license__ = "LGPLv3"


import sys
sys.path.insert(0,"/home/ble/buildAkantu/buildDebug/python")
print(sys.path)

import akantu as aka
import numpy as np
import matplotlib.pyplot as plt

from czm_damage import *
					
def set_dumpers(model):
    model.setBaseName("cohesive")
    model.addDumpFieldVector("displacement")
    model.addDumpFieldVector("external_force")
    model.addDumpField("strain")
    model.addDumpField("stress")
    model.addDumpField("blocked_dofs")

    model.setBaseNameToDumper("cohesive elements", "cohesive")
    model.addDumpFieldVectorToDumper("cohesive elements", "displacement")
    model.addDumpFieldToDumper("cohesive elements", "damage")
    model.addDumpFieldVectorToDumper("cohesive elements", "tractions")
    model.addDumpFieldVectorToDumper("cohesive elements", "opening")


def solve(material_file, mesh_file, increment):
    aka.parseInput(material_file)
    spatial_dimension = 2

    # -------------------------------------------------------------------------
    # Initialization
    # -------------------------------------------------------------------------
    mesh = aka.Mesh(spatial_dimension)
    mesh.read(mesh_file)

    model = aka.SolidMechanicsModelCohesive(mesh)
    model.getElementInserter().setLimit(aka._x, -0.1, 0.1);

    model.initFull(_analysis_method=aka._static, _is_extrinsic=False)
                           
    set_dumpers(model)

    L = 0.4
    E = model.getMaterial("steel").getReal("E")
    Gc = model.getMaterial("cohesive").getReal("G_c")
    tc = model.getMaterial("cohesive").getReal("sigma_c")
    wc = 2.*Gc/tc
    k = model.getMaterial("cohesive").getReal("k")
    g = degradation_function_linear_czm()
    h = softening_function_linear_czm(k,tc,Gc)
    updater = damage_updater(Gc,tc,k,g,h)

    imposed_disp = [0.]
    reaction_force = [0.]
    U = 0.
    F = 0.
        
    # -------------------------------------------------------------------------
    # Boundary conditions
    # -------------------------------------------------------------------------
    model.applyBC(aka.FixedValue(0.0, aka._x), "left")
    model.applyBC(aka.FixedValue(0.0, aka._y), "point")
    model.applyBC(aka.IncrementValue(increment, aka._x), "right")  
    U+=increment
    imposed_disp.append(U)
    
    model.getExternalForce()[:] = 0

    solver = model.getNonLinearSolver("static")
    solver.set("max_iterations", 2)
    solver.set("threshold", 1e-10)
    solver.set("convergence_type", aka.SolveConvergenceCriteria.residual)

    model.getNewSolver("linear_static", aka.TimeStepSolverType.static,
                       aka.NonLinearSolverType.linear)
    model.setIntegrationScheme("linear_static", "displacement",
                       aka.IntegrationSchemeType.pseudo_time)
    model.setIntegrationScheme("linear_static", "lambda",
                       aka.IntegrationSchemeType.pseudo_time)
                                                  
    d = model.getMaterial("cohesive").getInternalReal("czm_damage")(aka._cohesive_2d_4)
    lda = model.getMaterial("cohesive").getInternalReal("lambda")(aka._cohesive_2d_4)
    Fint = model.getInternalForce() 
    
    nodes_right = mesh.getElementGroup("right").getNodeGroup().getNodes()
    model.solveStep("linear_static")
    F = -np.sum(Fint[nodes_right,0])
    reaction_force.append(F)
    
    model.dump()
    #model.dump("cohesive elements")
            
    maxsteps = 8
    maxiter = 10
    tol = 1e-6
    
    for i in range(0, maxsteps):
        print("---- step {0}/{1}".format(i, maxsteps))
        model.applyBC(aka.IncrementValue(increment, aka._x), "right")
        U+=increment
        imposed_disp.append(U)    
        it = 0
        d_previous = model.getMaterial("cohesive").getInternalReal("czm_damage")(aka._cohesive_2d_4).copy()
        W0 = energy(model)
        while it < maxiter:
            #aka.debug.setDebugLevel(aka.dblInfo);
            model.solveStep("linear_static")
            d[:] = updater.update_d(d_previous,lda)
            W = energy(model)
            err = abs(W-W0)/W
            print("iter ",it,", err = ",err)          
            if err < tol:
                break
            else:
                W0 = W
                it+=1
        F = -np.sum(Fint[nodes_right,0])
        reaction_force.append(F)
    
        model.dump()
        #model.dump("cohesive elements")

    plt.plot(imposed_disp,reaction_force,'-o')
    plt.plot([0., (tc/E)*L, wc], [0., tc*L, 0.], 'k-o')
    plt.show()
    W = model.getEnergy("potential");
    D = model.getEnergy("dissipated");
    print("Elastic energy = ",W)
    print("Dissipated energy = ",D)

# -----------------------------------------------------------------------------
# main
# -----------------------------------------------------------------------------
def main():
    mesh_file = "triangle.msh"
    material_file = "material.dat"
    increment = 0.01e-3
    solve(material_file, mesh_file, increment)


# -----------------------------------------------------------------------------
if __name__ == "__main__":
    main()
